src/server/database/votes.ts

```diff
@@ -112,33 +119,78 @@ export default class Votes {
         } catch (error) {
             return Promise.reject(error);
         }
     };
 
-    addVoteBatch = async (votes: Vote[]): Promise<number> => {
+    addVoteBatch = async (votes: Vote[]): Promise<BatchUploadResults> => {
         try {
+            const output: BatchUploadResults = new BatchUploadResults();
             // Find Marosijo user_client or create a new one if none exists
             const marosijo_client_id = await this.findMarosijoClientId();
+            const validatedIds = await this.getValidatedClips(votes);
+
+            // Store for output
+            output.foundValidatedClips = validatedIds.length;
+
+            // Filter out clips that have been validated
+            const toAdd = votes.filter((vote) => {
+                return !validatedIds.some(
+                    (validated) => validated == vote.clipId
+                );
+            });
+
+            // Create an array for the VALUES that are to be added to the votes table
+            const values: string[] = Array<string>();
+            toAdd.forEach((vote) => {
+                const unsure = ClipVote.UNSURE == vote.vote;
+                const isValid = unsure
+                    ? null
+                    : vote.vote == ClipVote.VALID
+                    ? true
+                    : false;
+                values.push(
+                    `(null, ${vote.clipId}, '${marosijo_client_id}', ${isValid}, ${vote.isSuper}, ${unsure} )`
+                );
+            });
+
+            // Add votes to votes table
+            const res = await this.sql.query(`
+                    INSERT INTO
+                        votes (id, clip_id, client_id, is_valid, is_super, is_unsure)
+                    VALUES
+                        ${values.join(',')}
+                    ON DUPLICATE KEY UPDATE
+                        is_valid = VALUES(is_valid),
+                        is_unsure = VALUES(is_unsure)
+
+                        `);
+            output.duplicateVotes = this.getDuplicates(res[0]);
+            output.votesAdded = this.getNewRows(res[0]);
+
+            // Update clips with super votes
+            const supers = toAdd.filter((vote) => vote.isSuper);
+            const clipsUpdatedSuper = await this.updateClipsFromMachineBatch(
+                supers
+            );
 
-            // Map the votes to db inserts
-            const results = await Promise.all(
-                votes.map(async (vote) => {
-                    const saveVoteResult = await this.saveVote(
-                        marosijo_client_id,
-                        vote.clipId,
-                        false,
-                        vote.vote
-                    );
-                    return Promise.resolve(saveVoteResult);
-                })
+            // Update clips if needed where vote is not supervote
+            const normalVotes = toAdd.filter((vote) => !vote.isSuper);
+            const clipsUpdatedNormal = await this.updateFromBatchIfNeeded(
```

Comment:
Or maybe that is taken care of here? :P 
Didn't see this until afterwards.

But what is happening in line 163-173 then?  There it says "Add votes to votes table" but here you seem to be updating votes as well? Are you perhaps updating the clips table according to the newly added votes to the vote table?

Reply:
Yes! Here we are only updating the clips table with **is_valid** if we can count up to 2 valid or 2 invalid votes in the votes table for the clip.